File: src/test/java/io/proleap/cobol/e2e/SaleadsMiNegocioFullTest.java
package io.proleap.cobol.e2e;

import static org.junit.Assert.fail;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.junit.Assume;
import org.junit.Test;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserContext;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.Playwright;
import com.microsoft.playwright.PlaywrightException;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.WaitForSelectorState;

/**
 * Full E2E workflow for SaleADS "Mi Negocio" module.
 *
 * Runtime configuration:
 * - SALEADS_LOGIN_URL: login page URL for any target environment (required).
 * - SALEADS_HEADLESS: "true" or "false" (optional, default true).
 *
 * This test deliberately avoids hardcoded domains and uses visible-text selectors whenever possible.
 */
public class SaleadsMiNegocioFullTest {

	private static final String GOOGLE_ACCOUNT_EMAIL = "juanlucasbarbiergarzon@gmail.com";
	private static final String TEST_NAME = "saleads_mi_negocio_full_test";

	@Test
	public void saleadsMiNegocioWorkflow() throws IOException {
		final String loginUrl = System.getenv("SALEADS_LOGIN_URL");
		Assume.assumeTrue("Set SALEADS_LOGIN_URL to run this environment-agnostic E2E test.",
				loginUrl != null && !loginUrl.isBlank());

		final boolean headless = Boolean.parseBoolean(System.getenv().getOrDefault("SALEADS_HEADLESS", "true"));
		final Map<String, Boolean> finalReport = new LinkedHashMap<>();
		final List<String> failures = new ArrayList<>();
		final List<String> evidenceNotes = new ArrayList<>();
		final Path evidenceDir = buildEvidenceDir();

		try (Playwright playwright = Playwright.create()) {
			final Browser browser = playwright.chromium()
					.launch(new BrowserType.LaunchOptions().setHeadless(headless).setSlowMo(200));
			final BrowserContext context = browser.newContext();
			final Page appPage = context.newPage();

			appPage.navigate(loginUrl);
			waitForUiLoad(appPage);

			// Step 1: Login with Google
			final boolean loginClicked = clickVisibleText(appPage, "Sign in with Google", "Iniciar sesión con Google",
					"Continuar con Google", "Google");
			if (loginClicked) {
				waitForUiLoad(appPage);
				pickGoogleAccountIfVisible(context, GOOGLE_ACCOUNT_EMAIL);
				waitForUiLoad(appPage);
			}

			final boolean mainInterfaceVisible = isAnyTextVisible(appPage, "Negocio", "Mi Negocio", "Dashboard", "Inicio");
			final boolean leftSidebarVisible = isLeftSidebarVisible(appPage);
			captureScreenshot(appPage, evidenceDir.resolve("01-dashboard-loaded.png"), true);
			recordStep(finalReport, failures, "Login", mainInterfaceVisible && leftSidebarVisible,
					"Main interface or left sidebar not visible after Google login.");

			// Step 2: Open Mi Negocio menu
			clickVisibleText(appPage, "Negocio");
			clickVisibleText(appPage, "Mi Negocio");
			final boolean agregarNegocioVisibleInMenu = isAnyTextVisible(appPage, "Agregar Negocio");
			final boolean administrarNegociosVisibleInMenu = isAnyTextVisible(appPage, "Administrar Negocios");
			captureScreenshot(appPage, evidenceDir.resolve("02-mi-negocio-menu-expanded.png"), true);
			recordStep(finalReport, failures, "Mi Negocio menu",
					agregarNegocioVisibleInMenu && administrarNegociosVisibleInMenu,
					"Mi Negocio menu did not show expected submenu options.");

			// Step 3: Validate Agregar Negocio modal
			clickVisibleText(appPage, "Agregar Negocio");
			final boolean modalTitleVisible = waitForAnyTextVisible(appPage, 8000, "Crear Nuevo Negocio");
			final boolean nombreDelNegocioInputVisible = isAnyTextVisible(appPage, "Nombre del Negocio");
			final boolean businessQuotaVisible = isAnyTextVisible(appPage, "Tienes 2 de 3 negocios");
			final boolean modalButtonsVisible = isAnyTextVisible(appPage, "Cancelar")
					&& isAnyTextVisible(appPage, "Crear Negocio");
			captureScreenshot(appPage, evidenceDir.resolve("03-agregar-negocio-modal.png"), true);

			clickVisibleText(appPage, "Nombre del Negocio");
			fillIfInputVisible(appPage, "Negocio Prueba Automatización");
			clickVisibleText(appPage, "Cancelar");
			waitForUiLoad(appPage);

			recordStep(finalReport, failures, "Agregar Negocio modal",
					modalTitleVisible && nombreDelNegocioInputVisible && businessQuotaVisible && modalButtonsVisible,
					"Agregar Negocio modal did not match expected title/fields/buttons.");

			// Step 4: Open Administrar Negocios view
			clickVisibleText(appPage, "Mi Negocio");
			clickVisibleText(appPage, "Administrar Negocios");
			waitForUiLoad(appPage);
			final boolean infoGeneralSection = isAnyTextVisible(appPage, "Información General");
			final boolean detallesCuentaSection = isAnyTextVisible(appPage, "Detalles de la Cuenta");
			final boolean tusNegociosSection = isAnyTextVisible(appPage, "Tus Negocios");
			final boolean seccionLegalSection = isAnyTextVisible(appPage, "Sección Legal");
			captureScreenshot(appPage, evidenceDir.resolve("04-administrar-negocios-full-page.png"), true);
			recordStep(finalReport, failures, "Administrar Negocios view",
					infoGeneralSection && detallesCuentaSection && tusNegociosSection && seccionLegalSection,
					"Administrar Negocios sections are incomplete.");

			// Step 5: Validate Información General
			final boolean userNameVisible = isUserNameVisible(appPage);
			final boolean userEmailVisible = isUserEmailVisible(appPage);
			final boolean businessPlanVisible = isAnyTextVisible(appPage, "BUSINESS PLAN");
			final boolean cambiarPlanVisible = isAnyTextVisible(appPage, "Cambiar Plan");
			recordStep(finalReport, failures, "Información General",
					userNameVisible && userEmailVisible && businessPlanVisible && cambiarPlanVisible,
					"Información General did not include name/email/plan/actions.");

			// Step 6: Validate Detalles de la Cuenta
			final boolean cuentaCreadaVisible = isAnyTextVisible(appPage, "Cuenta creada");
			final boolean estadoActivoVisible = isAnyTextVisible(appPage, "Estado activo");
			final boolean idiomaSeleccionadoVisible = isAnyTextVisible(appPage, "Idioma seleccionado");
			recordStep(finalReport, failures, "Detalles de la Cuenta",
					cuentaCreadaVisible && estadoActivoVisible && idiomaSeleccionadoVisible,
					"Detalles de la Cuenta missing expected fields.");

			// Step 7: Validate Tus Negocios
			final boolean businessListVisible = isAnyTextVisible(appPage, "Tus Negocios");
			final boolean addBusinessButtonVisible = isAnyTextVisible(appPage, "Agregar Negocio");
			final boolean quotaVisibleInBusinessSection = isAnyTextVisible(appPage, "Tienes 2 de 3 negocios");
			recordStep(finalReport, failures, "Tus Negocios",
					businessListVisible && addBusinessButtonVisible && quotaVisibleInBusinessSection,
					"Tus Negocios section missing list/button/quota.");

			// Step 8: Validate Términos y Condiciones
			final LegalValidationResult termsResult = validateLegalLink(context, appPage, evidenceDir,
					"05-terminos-y-condiciones.png", "Términos y Condiciones", "Términos y Condiciones");
			evidenceNotes.add("Términos y Condiciones URL: " + termsResult.url);
			recordStep(finalReport, failures, "Términos y Condiciones", termsResult.passed,
					"Términos y Condiciones page heading/content validation failed.");

			// Step 9: Validate Política de Privacidad
			final LegalValidationResult privacyResult = validateLegalLink(context, appPage, evidenceDir,
					"06-politica-de-privacidad.png", "Política de Privacidad", "Política de Privacidad");
			evidenceNotes.add("Política de Privacidad URL: " + privacyResult.url);
			recordStep(finalReport, failures, "Política de Privacidad", privacyResult.passed,
					"Política de Privacidad page heading/content validation failed.");
		}

		writeFinalReport(evidenceDir, finalReport, evidenceNotes);

		if (!failures.isEmpty()) {
			fail("SaleADS Mi Negocio full workflow failures:\n - " + String.join("\n - ", failures));
		}
	}

	private static LegalValidationResult validateLegalLink(final BrowserContext context, final Page appPage,
			final Path evidenceDir, final String screenshotName, final String linkText, final String headingText) {
		final int previousPageCount = context.pages().size();
		final boolean clicked = clickVisibleText(appPage, linkText);
		if (!clicked) {
			return new LegalValidationResult(false, "N/A");
		}

		final Page legalPage = waitForPotentialNewTab(context, appPage, previousPageCount);
		waitForUiLoad(legalPage);

		final boolean headingVisible = isAnyTextVisible(legalPage, headingText);
		final boolean legalContentVisible = hasNonTrivialContent(legalPage);
		captureScreenshot(legalPage, evidenceDir.resolve(screenshotName), true);

		final String finalUrl = legalPage.url();
		if (legalPage != appPage) {
			legalPage.close();
			appPage.bringToFront();
			waitForUiLoad(appPage);
		}

		return new LegalValidationResult(headingVisible && legalContentVisible, finalUrl);
	}

	private static Page waitForPotentialNewTab(final BrowserContext context, final Page appPage,
			final int previousPageCount) {
		final long deadlineMs = System.currentTimeMillis() + 6000;
		while (System.currentTimeMillis() < deadlineMs) {
			if (context.pages().size() > previousPageCount) {
				return context.pages().get(context.pages().size() - 1);
			}
			appPage.waitForTimeout(250);
		}
		return appPage;
	}

	private static void pickGoogleAccountIfVisible(final BrowserContext context, final String email) {
		final long deadlineMs = System.currentTimeMillis() + 20000;
		while (System.currentTimeMillis() < deadlineMs) {
			for (final Page page : context.pages()) {
				final Locator emailOption = page.getByText(email, new Page.GetByTextOptions().setExact(false)).first();
				if (isVisible(emailOption, 1000)) {
					emailOption.click();
					waitForUiLoad(page);
					return;
				}
			}

			for (final Page page : context.pages()) {
				if (isAnyTextVisible(page, "Choose an account", "Elige una cuenta")) {
					page.waitForTimeout(300);
				}
			}
			context.pages().get(0).waitForTimeout(300);
		}
	}

	private static boolean isUserNameVisible(final Page page) {
		final Locator textNodes = page.locator("h1,h2,h3,p,span,div");
		try {
			final int sample = Math.min(textNodes.count(), 40);
			for (int i = 0; i < sample; i++) {
				final String text = textNodes.nth(i).innerText().trim();
				if (!text.isEmpty() && !text.contains("@") && text.split("\\s+").length >= 2
						&& text.length() >= 5 && !text.equalsIgnoreCase("Información General")) {
					return true;
				}
			}
		} catch (final PlaywrightException ignored) {
			return false;
		}
		return false;
	}

	private static boolean isUserEmailVisible(final Page page) {
		return hasVisibleRegexLikeText(page, "@", ".");
	}

	private static boolean hasVisibleRegexLikeText(final Page page, final String mandatoryPart,
			final String secondaryPart) {
		final Locator textNodes = page.locator("p,span,div,a");
		try {
			final int sample = Math.min(textNodes.count(), 120);
			for (int i = 0; i < sample; i++) {
				final String text = textNodes.nth(i).innerText().trim();
				if (!text.isEmpty() && text.contains(mandatoryPart) && text.contains(secondaryPart)) {
					return true;
				}
			}
		} catch (final PlaywrightException ignored) {
			return false;
		}
		return false;
	}

	private static boolean hasNonTrivialContent(final Page page) {
		try {
			final String text = page.locator("body").innerText();
			return text != null && text.trim().length() >= 120;
		} catch (final PlaywrightException ignored) {
			return false;
		}
	}

	private static boolean isLeftSidebarVisible(final Page page) {
		try {
			final Locator sidebar = page.locator("aside, nav").first();
			return isVisible(sidebar, 6000);
		} catch (final PlaywrightException ignored) {
			return false;
		}
	}

	private static void fillIfInputVisible(final Page page, final String value) {
		final Locator input = page.locator("input").first();
		if (isVisible(input, 2500)) {
			input.fill(value);
		}
	}

	private static boolean clickVisibleText(final Page page, final String... labels) {
		for (final String label : labels) {
			final Locator candidate = page.getByText(label, new Page.GetByTextOptions().setExact(false)).first();
			if (isVisible(candidate, 2500)) {
				candidate.click();
				waitForUiLoad(page);
				return true;
			}
		}
		return false;
	}

	private static boolean waitForAnyTextVisible(final Page page, final int timeoutMs, final String... labels) {
		final long deadlineMs = System.currentTimeMillis() + timeoutMs;
		while (System.currentTimeMillis() < deadlineMs) {
			if (isAnyTextVisible(page, labels)) {
				return true;
			}
			page.waitForTimeout(200);
		}
		return false;
	}

	private static boolean isAnyTextVisible(final Page page, final String... labels) {
		for (final String label : labels) {
			final Locator candidate = page.getByText(label, new Page.GetByTextOptions().setExact(false)).first();
			if (isVisible(candidate, 1800)) {
				return true;
			}
		}
		return false;
	}

	private static boolean isVisible(final Locator locator, final int timeoutMs) {
		try {
			locator.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE).setTimeout(timeoutMs));
			return locator.isVisible();
		} catch (final PlaywrightException ignored) {
			return false;
		}
	}

	private static void waitForUiLoad(final Page page) {
		try {
			page.waitForLoadState(LoadState.DOMCONTENTLOADED);
		} catch (final PlaywrightException ignored) {
			// ignore transient load-state issues and continue with explicit waits by visible text
		}

		try {
			page.waitForLoadState(LoadState.NETWORKIDLE);
		} catch (final PlaywrightException ignored) {
			// some pages continuously poll; continue after a bounded delay
		}

		page.waitForTimeout(700);
	}

	private static void captureScreenshot(final Page page, final Path path, final boolean fullPage) {
		try {
			page.screenshot(new Page.ScreenshotOptions().setPath(path).setFullPage(fullPage));
		} catch (final PlaywrightException ignored) {
			// best effort evidence capture
		}
	}

	private static Path buildEvidenceDir() throws IOException {
		final String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
		final Path path = Path.of("target", "evidence", TEST_NAME, timestamp);
		Files.createDirectories(path);
		return path;
	}

	private static void writeFinalReport(final Path evidenceDir, final Map<String, Boolean> finalReport,
			final List<String> evidenceNotes) throws IOException {
		final StringBuilder sb = new StringBuilder();
		sb.append("Test: ").append(TEST_NAME).append('\n');
		sb.append("Generated at: ").append(LocalDateTime.now()).append('\n');
		sb.append('\n');
		sb.append("Final Report (PASS/FAIL):").append('\n');
		for (final Map.Entry<String, Boolean> entry : finalReport.entrySet()) {
			sb.append("- ").append(entry.getKey()).append(": ").append(entry.getValue() ? "PASS" : "FAIL").append('\n');
		}

		if (!evidenceNotes.isEmpty()) {
			sb.append('\n').append("Evidence Notes:").append('\n');
			for (final String note : evidenceNotes) {
				sb.append("- ").append(note).append('\n');
			}
		}

		Files.writeString(evidenceDir.resolve("final-report.txt"), sb.toString());
		System.out.println(sb);
	}

	private static void recordStep(final Map<String, Boolean> finalReport, final List<String> failures, final String key,
			final boolean passed, final String failureMessage) {
		finalReport.put(key, passed);
		if (!passed) {
			failures.add(key + ": " + failureMessage);
		}
	}

	private static final class LegalValidationResult {
		private final boolean passed;
		private final String url;

		private LegalValidationResult(final boolean passed, final String url) {
			this.passed = passed;
			this.url = url;
		}
	}
}

File: automation/saleads-mi-negocio/scripts/saleads_mi_negocio_full_test.mjs
#!/usr/bin/env node

import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { chromium } from "playwright";

const TEST_NAME = "saleads_mi_negocio_full_test";
const GOOGLE_ACCOUNT = process.env.SALEADS_GOOGLE_ACCOUNT ?? "juanlucasbarbiergarzon@gmail.com";
const LOGIN_URL = process.env.SALEADS_LOGIN_URL;
const HEADLESS = process.env.SALEADS_HEADLESS !== "false";
const WAIT_AFTER_CLICK_MS = Number.parseInt(process.env.SALEADS_WAIT_MS ?? "1200", 10);
const ARTIFACTS_DIR = path.resolve(
  process.env.SALEADS_ARTIFACTS_DIR ?? "./artifacts/saleads-mi-negocio",
);

const REPORT_FIELDS = [
  "Login",
  "Mi Negocio menu",
  "Agregar Negocio modal",
  "Administrar Negocios view",
  "Información General",
  "Detalles de la Cuenta",
  "Tus Negocios",
  "Términos y Condiciones",
  "Política de Privacidad",
];

function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function slugify(value) {
  return value
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-zA-Z0-9]+/g, "-")
    .replace(/(^-|-$)/g, "")
    .toLowerCase();
}

function timestamp() {
  return new Date().toISOString().replace(/[:.]/g, "-");
}

async function waitForUi(page) {
  await page.waitForLoadState("domcontentloaded", { timeout: 20000 }).catch(() => {});
  await page.waitForLoadState("networkidle", { timeout: 10000 }).catch(() => {});
  await page.waitForTimeout(WAIT_AFTER_CLICK_MS);
}

function buildTextLocators(page, text) {
  const nameRegex = new RegExp(escapeRegExp(text), "i");
  return [
    page.getByRole("button", { name: nameRegex }).first(),
    page.getByRole("link", { name: nameRegex }).first(),
    page.getByRole("menuitem", { name: nameRegex }).first(),
    page.getByRole("tab", { name: nameRegex }).first(),
    page.getByText(nameRegex).first(),
  ];
}

async function findVisibleLocatorByTexts(page, texts, timeoutMs = 15000) {
  const deadline = Date.now() + timeoutMs;

  while (Date.now() < deadline) {
    for (const text of texts) {
      for (const locator of buildTextLocators(page, text)) {
        const visible = await locator.isVisible().catch(() => false);
        if (visible) {
          return { locator, matchedText: text };
        }
      }
    }
    await page.waitForTimeout(250);
  }

  throw new Error(`None of these texts were visible: ${texts.join(" | ")}`);
}

async function clickByTexts(page, texts, options = {}) {
  const { timeoutMs = 20000, waitAfterClick = true } = options;
  const { locator, matchedText } = await findVisibleLocatorByTexts(page, texts, timeoutMs);
  await locator.click();
  if (waitAfterClick) {
    await waitForUi(page);
  }
  return matchedText;
}

async function assertTextsVisible(page, texts, timeoutMs = 15000) {
  for (const text of texts) {
    const { locator } = await findVisibleLocatorByTexts(page, [text], timeoutMs);
    const visible = await locator.isVisible().catch(() => false);
    if (!visible) {
      throw new Error(`Expected visible text not found: ${text}`);
    }
  }
}

async function assertAnyTextVisible(page, texts, timeoutMs = 15000) {
  await findVisibleLocatorByTexts(page, texts, timeoutMs);
}

async function screenshot(page, label, options = {}) {
  const filename = `${timestamp()}-${slugify(label)}.png`;
  const outputPath = path.join(ARTIFACTS_DIR, filename);
  await page.screenshot({ path: outputPath, fullPage: options.fullPage ?? false });
  return outputPath;
}

function createResults() {
  return Object.fromEntries(
    REPORT_FIELDS.map((field) => [field, { status: "FAIL", details: [], evidence: {} }]),
  );
}

function createReport() {
  return {
    name: TEST_NAME,
    started_at: new Date().toISOString(),
    finished_at: null,
    config: {
      login_url: LOGIN_URL ?? null,
      google_account: GOOGLE_ACCOUNT,
      headless: HEADLESS,
      wait_after_click_ms: WAIT_AFTER_CLICK_MS,
      artifacts_dir: ARTIFACTS_DIR,
    },
    results: createResults(),
    summary: {
      passed: 0,
      failed: 0,
      status: "FAIL",
    },
  };
}

function markStep(report, step, status, details, evidence = {}) {
  report.results[step] = {
    status,
    details: Array.isArray(details) ? details : [details],
    evidence,
  };
}

function computeSummary(report) {
  let passed = 0;
  let failed = 0;

  for (const field of REPORT_FIELDS) {
    if (report.results[field].status === "PASS") {
      passed += 1;
    } else {
      failed += 1;
    }
  }

  report.summary = {
    passed,
    failed,
    status: failed === 0 ? "PASS" : "FAIL",
  };
  report.finished_at = new Date().toISOString();
}

async function trySelectGoogleAccount(page) {
  const accountLocator = page.getByText(new RegExp(`^\\s*${escapeRegExp(GOOGLE_ACCOUNT)}\\s*$`, "i"));
  const visible = await accountLocator.first().isVisible().catch(() => false);
  if (visible) {
    await accountLocator.first().click();
    await waitForUi(page);
    return true;
  }

  return false;
}

async function openMiNegocioMenu(page) {
  const submenuTexts = ["Agregar Negocio", "Administrar Negocios"];
  const submenuReady = await isAnyTextVisible(page, submenuTexts, 2000);
  if (submenuReady) {
    return;
  }

  const menuAttempts = [
    ["Mi Negocio"],
    ["Negocio", "Mi Negocio"],
  ];

  for (const attempt of menuAttempts) {
    try {
      for (const text of attempt) {
        await clickByTexts(page, [text], { timeoutMs: 10000 });
      }
      const expanded = await isAnyTextVisible(page, submenuTexts, 5000);
      if (expanded) {
        return;
      }
    } catch {
      // Try the next menu strategy.
    }
  }

  throw new Error("Unable to expand Mi Negocio menu.");
}

async function isAnyTextVisible(page, texts, timeoutMs = 3000) {
  try {
    await assertAnyTextVisible(page, texts, timeoutMs);
    return true;
  } catch {
    return false;
  }
}

function hasNameLikeText(sectionText) {
  const lines = sectionText
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean)
    .filter((line) => !/@/.test(line))
    .filter((line) => !/informaci[oó]n general|business plan|cambiar plan/i.test(line));

  return lines.some((line) => /^[A-Za-zÀ-ÿ' -]{4,}$/.test(line));
}

async function validateLegalLink(context, appPage, linkText, headingText, label) {
  const appUrlBeforeClick = appPage.url();
  const popupPromise = context.waitForEvent("page", { timeout: 7000 }).catch(() => null);
  await clickByTexts(appPage, [linkText], { waitAfterClick: false });
  const popup = await popupPromise;

  const legalPage = popup ?? appPage;
  await waitForUi(legalPage);

  await assertAnyTextVisible(legalPage, [headingText], 25000);
  const bodyText = await legalPage.locator("body").innerText();
  if (bodyText.trim().length < 100) {
    throw new Error(`Legal page "${headingText}" loaded but visible content looks too short.`);
  }

  const capture = await screenshot(legalPage, label, { fullPage: true });
  const finalUrl = legalPage.url();

  if (popup) {
    await popup.close();
    await appPage.bringToFront();
    await waitForUi(appPage);
  } else if (appPage.url() !== appUrlBeforeClick) {
    await appPage.goBack().catch(() => {});
    await waitForUi(appPage);
  }

  return { screenshot: capture, final_url: finalUrl };
}

async function findBusinessNameInput(page) {
  const labelLocator = page.getByLabel(/Nombre del Negocio/i).first();
  if (await labelLocator.isVisible().catch(() => false)) {
    return labelLocator;
  }

  const placeholderLocator = page.getByPlaceholder(/Nombre del Negocio/i).first();
  if (await placeholderLocator.isVisible().catch(() => false)) {
    return placeholderLocator;
  }

  const genericInput = page.locator("input").first();
  if (await genericInput.isVisible().catch(() => false)) {
    return genericInput;
  }

  throw new Error("Could not locate the input field 'Nombre del Negocio'.");
}

async function run() {
  const report = createReport();
  await fs.mkdir(ARTIFACTS_DIR, { recursive: true });
  let runnerError = null;

  const browser = await chromium.launch({
    headless: HEADLESS,
  });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 },
  });
  const page = await context.newPage();

  try {
    if (!LOGIN_URL) {
      throw new Error(
        "SALEADS_LOGIN_URL is required in this runtime. No URL is hardcoded by design to support any environment.",
      );
    }

    await page.goto(LOGIN_URL, { waitUntil: "domcontentloaded" });
    await waitForUi(page);

    // Step 1 - Login with Google.
    try {
      const popupPromise = context.waitForEvent("page", { timeout: 7000 }).catch(() => null);
      await clickByTexts(page, [
        "Sign in with Google",
        "Iniciar sesión con Google",
        "Inicia sesión con Google",
        "Continuar con Google",
        "Login with Google",
      ]);
      const popup = await popupPromise;
      const authPage = popup ?? page;
      await waitForUi(authPage);

      await trySelectGoogleAccount(authPage);

      if (popup) {
        await popup.waitForEvent("close", { timeout: 45000 }).catch(() => {});
        await page.bringToFront();
      }

      await waitForUi(page);
      await assertAnyTextVisible(page, ["Mi Negocio", "Negocio"], 45000);

      const sidebarVisible =
        (await page.locator("aside").first().isVisible().catch(() => false)) ||
        (await page.getByRole("navigation").first().isVisible().catch(() => false));
      if (!sidebarVisible) {
        throw new Error("Main application loaded but sidebar navigation is not visible.");
      }

      const dashShot = await screenshot(page, "dashboard-loaded");
      markStep(
        report,
        "Login",
        "PASS",
        [
          "Main application interface is visible after Google sign-in.",
          "Left sidebar navigation is visible.",
        ],
        { screenshot: dashShot },
      );
    } catch (error) {
      const failShot = await screenshot(page, "login-failure").catch(() => null);
      markStep(report, "Login", "FAIL", error.message, { screenshot: failShot });
    }

    // Step 2 - Open Mi Negocio menu.
    try {
      await openMiNegocioMenu(page);
      await assertTextsVisible(page, ["Agregar Negocio", "Administrar Negocios"], 10000);
      const menuShot = await screenshot(page, "mi-negocio-menu-expanded");
      markStep(
        report,
        "Mi Negocio menu",
        "PASS",
        [
          "Mi Negocio submenu expanded successfully.",
          "'Agregar Negocio' and 'Administrar Negocios' are visible.",
        ],
        { screenshot: menuShot },
      );
    } catch (error) {
      const failShot = await screenshot(page, "mi-negocio-menu-failure").catch(() => null);
      markStep(report, "Mi Negocio menu", "FAIL", error.message, { screenshot: failShot });
    }

    // Step 3 - Validate Agregar Negocio modal.
    try {
      await openMiNegocioMenu(page);
      await clickByTexts(page, ["Agregar Negocio"]);
      await assertTextsVisible(
        page,
        ["Crear Nuevo Negocio", "Nombre del Negocio", "Tienes 2 de 3 negocios", "Cancelar", "Crear Negocio"],
        15000,
      );

      const input = await findBusinessNameInput(page);
      await input.click();
      await input.fill("Negocio Prueba Automatización");
      const modalShot = await screenshot(page, "agregar-negocio-modal");
      await clickByTexts(page, ["Cancelar"]);
      markStep(
        report,
        "Agregar Negocio modal",
        "PASS",
        [
          "Modal 'Crear Nuevo Negocio' opened with required input, quota, and action buttons.",
          "Typed sample business name and closed with 'Cancelar'.",
        ],
        { screenshot: modalShot },
      );
    } catch (error) {
      const failShot = await screenshot(page, "agregar-negocio-modal-failure").catch(() => null);
      markStep(report, "Agregar Negocio modal", "FAIL", error.message, { screenshot: failShot });
    }

    // Step 4 - Open Administrar Negocios.
    try {
      await openMiNegocioMenu(page);
      await clickByTexts(page, ["Administrar Negocios"]);
      await assertTextsVisible(
        page,
        ["Información General", "Detalles de la Cuenta", "Tus Negocios", "Sección Legal"],
        20000,
      );
      const accountShot = await screenshot(page, "administrar-negocios", { fullPage: true });
      markStep(
        report,
        "Administrar Negocios view",
        "PASS",
        [
          "Account page loaded with all expected sections.",
          "Información General, Detalles de la Cuenta, Tus Negocios, and Sección Legal are visible.",
        ],
        { screenshot: accountShot },
      );
    } catch (error) {
      const failShot = await screenshot(page, "administrar-negocios-failure").catch(() => null);
      markStep(report, "Administrar Negocios view", "FAIL", error.message, { screenshot: failShot });
    }

    // Step 5 - Validate Información General.
    try {
      await assertTextsVisible(page, ["Información General", "BUSINESS PLAN", "Cambiar Plan"], 12000);
      const infoGeneralText = await page.locator("body").innerText();
      const emailVisible = /[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i.test(infoGeneralText);
      if (!emailVisible) {
        throw new Error("No user email address was detected on the page.");
      }
      if (!hasNameLikeText(infoGeneralText)) {
        throw new Error("No user name-like text was detected on the page.");
      }

      markStep(
        report,
        "Información General",
        "PASS",
        [
          "User name-like text is visible.",
          "User email is visible.",
          "BUSINESS PLAN and 'Cambiar Plan' are visible.",
        ],
      );
    } catch (error) {
      markStep(report, "Información General", "FAIL", error.message);
    }

    // Step 6 - Validate Detalles de la Cuenta.
    try {
      await assertTextsVisible(page, ["Cuenta creada", "Estado activo", "Idioma seleccionado"], 12000);
      markStep(
        report,
        "Detalles de la Cuenta",
        "PASS",
        [
          "'Cuenta creada' is visible.",
          "'Estado activo' is visible.",
          "'Idioma seleccionado' is visible.",
        ],
      );
    } catch (error) {
      markStep(report, "Detalles de la Cuenta", "FAIL", error.message);
    }

    // Step 7 - Validate Tus Negocios.
    try {
      await assertTextsVisible(page, ["Tus Negocios", "Agregar Negocio", "Tienes 2 de 3 negocios"], 12000);
      markStep(
        report,
        "Tus Negocios",
        "PASS",
        [
          "Business list section is visible.",
          "Agregar Negocio button exists.",
          "Business quota text is visible.",
        ],
      );
    } catch (error) {
      markStep(report, "Tus Negocios", "FAIL", error.message);
    }

    // Step 8 - Validate Términos y Condiciones.
    try {
      const legalResult = await validateLegalLink(
        context,
        page,
        "Términos y Condiciones",
        "Términos y Condiciones",
        "terminos-y-condiciones",
      );
      markStep(
        report,
        "Términos y Condiciones",
        "PASS",
        [
          "Legal page heading and body content are visible.",
          `Final URL: ${legalResult.final_url}`,
        ],
        legalResult,
      );
    } catch (error) {
      const failShot = await screenshot(page, "terminos-failure").catch(() => null);
      markStep(report, "Términos y Condiciones", "FAIL", error.message, { screenshot: failShot });
    }

    // Step 9 - Validate Política de Privacidad.
    try {
      const legalResult = await validateLegalLink(
        context,
        page,
        "Política de Privacidad",
        "Política de Privacidad",
        "politica-de-privacidad",
      );
      markStep(
        report,
        "Política de Privacidad",
        "PASS",
        [
          "Privacy page heading and body content are visible.",
          `Final URL: ${legalResult.final_url}`,
        ],
        legalResult,
      );
    } catch (error) {
      const failShot = await screenshot(page, "privacidad-failure").catch(() => null);
      markStep(report, "Política de Privacidad", "FAIL", error.message, { screenshot: failShot });
    }
  } catch (error) {
    runnerError = error;
    if (!report.results.Login.details.length) {
      markStep(report, "Login", "FAIL", error.message);
    }
  } finally {
    computeSummary(report);

    const reportPath = path.join(ARTIFACTS_DIR, `${timestamp()}-final-report.json`);
    await fs.writeFile(reportPath, `${JSON.stringify(report, null, 2)}\n`, "utf8");

    console.log("=== SaleADS Mi Negocio Full Test Report ===");
    for (const field of REPORT_FIELDS) {
      console.log(`${field}: ${report.results[field].status}`);
    }
    console.log(`Summary: ${report.summary.status} (passed: ${report.summary.passed}, failed: ${report.summary.failed})`);
    console.log(`Artifacts directory: ${ARTIFACTS_DIR}`);
    console.log(`Report file: ${reportPath}`);
    if (runnerError) {
      console.log(`Runner setup/runtime error: ${runnerError.message}`);
    }

    await context.close().catch(() => {});
    await browser.close().catch(() => {});

    process.exitCode = report.summary.failed === 0 ? 0 : 1;
  }
}

run().catch((error) => {
  console.error("Fatal test runner error:", error);
  process.exit(1);
});
